src/qibo/tomography/gate_set_tomography.py

```diff
@@ -141,13 +145,144 @@ def _measurement_basis(j, nqubits):
 
 
+def _extract_nqubits(gate, params=None):
+    """A function to extract the number of qubits the gate acts on.
+    Args:
+        gate (:class:`qibo.gates.abstract.Gate`): gate
+        params (list, optional): A list containing the angles for the gate.
+    Returns:
+        nqubits (int): Number of qubits that the gate acts on.
+    """
+
+    init_args = signature(gate).parameters
+    if "unitary" in init_args and params is not None:
+        nqubits = int(np.log2(np.shape(params[0])[0]))
+    else:
+        if "q" in init_args:
+            nqubits = 1
+        elif "q0" in init_args and "q1" in init_args and "q2" not in init_args:
+            nqubits = 2
+        else:
+            nqubits = None
+            raise_error(
+                RuntimeError,
+                f"Gate {gate} is not supported for `GST`, only 1- and 2-qubit gates are supported.",
+            )
+    return nqubits
+
+
+def _get_nqubits_and_angles(
+    gate: Union[gates.abstract.Gate, Tuple[gates.abstract.Gate, List[float]]],
+):
+    """A function to extract information about a `qibo.gates.Gate`.
+
+    Args:
+        gate (:class:`qibo.gates.abstract.Gate` or tuple): Either a gate or a tuple consisting of a gate and a list of its parameters.
+            Examples of a valid input:
+            - ``gate = gates.Z`` for a non-parametrized gate.
+            - ``gate = (gates.RX, [np.pi/3])`` or ``gate = (gates.PRX, [np.pi/2, np.pi/3])`` for a parametrized gate.
+            - ``gate = (gates.Unitary, [np.array([[1, 0], [0, 1]])])`` for an arbitrary unitary operator.
+    Returns:
+        gate (:class:`qibo.gates.Gate`): Gate class.
+        nqubits (int): Number of qubits that the gate acts on.
+        angle_names (list[str]): If gate is a parametrized gate, ``angle_names`` contains a list containing the angle names of the
+            parametrized gate. Else, ``None``.
+        angle_values (dict[str, float]): If gate is a parametrized gate, ``angle_values`` is a dictionary containing the angle names
+            of the parametrized gate and the respective angles. Else, an empty dictionary is returned.
+        params (list[float]): Stores all the parameters of the gate in a list.
+    """
+
+    if isinstance(gate, tuple):
+        angles = ANGLES
+        gate, params = gate
+        if not (isinstance(params, list) or isinstance(params, tuple)):
+            params = [params]
+    else:
+        angles = None
+        params = None
+    init_args = signature(gate).parameters
+    nqubits = _extract_nqubits(gate, params)
+
+    if angles is not None:
+        angle_names = [arg for arg in init_args if arg in angles]
+        angle_values = dict(zip(angle_names, params))
+    else:
+        angle_names = None
+        angle_values = {}
+
+    return gate, nqubits, angle_names, angle_values, params
+
+
+def _extract_gate(
+    gate: Union[gates.abstract.Gate, Tuple[gates.abstract.Gate, List[float]]],
+    idx: Optional[Union[int, Tuple[int, ...]]] = None,
```

Comment:
```suggestion
    qubits: Optional[Tuple[int, ...]] = None,
```
this is not a very explanatory name, just call it qubits and always ask for a tuple